File: android/EGETrainerApp/core/src/main/kotlin/ru/egetrainer/core/services/StudentService.kt
package ru.egetrainer.core.services

import kotlinx.serialization.builtins.ListSerializer
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.add
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import kotlinx.serialization.json.putJsonArray
import kotlinx.serialization.json.putJsonObject
import kotlinx.serialization.json.JsonObject
import ru.egetrainer.core.models.AnalyticsScreen
import ru.egetrainer.core.models.AttemptQuestion
import ru.egetrainer.core.models.IncomingTeacherRequest
import ru.egetrainer.core.models.MyTeacher
import ru.egetrainer.core.models.PickingScreen
import ru.egetrainer.core.models.ResolveBatchResult
import ru.egetrainer.core.models.ResolvedQuestion
import java.time.Instant
import java.util.UUID

/** Студенческая аналитика + consent-flow ученика + запись попыток тренировки (порт StudentService.swift). */
class StudentService(private val client: SupabaseClient) {

    /**
     * student_analytics_screen_v1 — canonical layer-4 контракт статистики.
     * scope: "self" (ученик о себе) | "teacher" (учитель об ученике, требуется studentId).
     */
    suspend fun analytics(
        scope: String = "self",
        studentId: String? = null,
        days: Int = 30,
        source: String = "all",
    ): AnalyticsScreen = client.rpc(
        "student_analytics_screen_v1",
        buildJsonObject {
            put("p_viewer_scope", scope)
            put("p_days", days)
            put("p_source", source)
            put("p_mode", "init")
            if (studentId != null) put("p_student_id", studentId)
        },
        AnalyticsScreen.serializer(),
    )

    /** id текущего пользователя — для self-гейта teacher RPC. */
    fun selfUserId(): String? = client.currentSession?.user?.id

    /**
     * Подбор задач с фильтром — self-гейт teacher_picking_resolve_batch_v1
     * (тот же RPC, p_student_id = свой id; как student-ветки picker.js).
     * requests: n уже с over-fetch.
     */
    suspend fun resolveFiltered(
        requests: List<ResolveRequest>,
        filterId: String?,
        excludeQuestionIds: List<String>,
        seed: String?,
    ): List<ResolvedQuestion> {
        val sid = selfUserId() ?: throw SupabaseError.AuthRequired()
        val result = client.rpc(
            "teacher_picking_resolve_batch_v1",
            buildJsonObject {
                put("p_student_id", sid)
                put("p_source", "all")
                if (filterId != null) put("p_filter_id", filterId) else put("p_filter_id", JsonNull)
                putJsonObject("p_selection") {}
                putJsonArray("p_requests") {
                    requests.forEach { r ->
                        add(buildJsonObject {
                            put("scope_kind", r.scopeKind)
                            put("scope_id", r.scopeId)
                            put("n", r.n)
                        })
                    }
                }
                if (seed != null) put("p_seed", seed) else put("p_seed", JsonNull)
                putJsonArray("p_exclude_question_ids") {
                    excludeQuestionIds.forEach { add(it) }
                }
            },
            ResolveBatchResult.serializer(),
        )
        return result.pickedQuestions.orEmpty()
    }

    /**
     * Экран подбора с фильтром для самого ученика (self-гейт
     * teacher_picking_screen_v2 — бейджи состояний тем как у учителя).
     */
    suspend fun pickingScreenSelf(filterId: String?, days: Int = 30): PickingScreen {
        val sid = selfUserId() ?: throw SupabaseError.AuthRequired()
        return client.rpc(
            "teacher_picking_screen_v2",
            buildJsonObject {
                put("p_student_id", sid)
                put("p_mode", "init")
                put("p_days", days)
                put("p_source", "all")
                if (filterId != null) put("p_filter_id", filterId) else put("p_filter_id", JsonNull)
                putJsonObject("p_selection") {}
                putJsonObject("p_request") {}
                put("p_seed", JsonNull)
                put("p_exclude_question_ids", JsonNull)
            },
            PickingScreen.serializer(),
        )
    }

    // MARK: Тренировка: запись попытки (write_answer_events_v1, supabase-write.js)

    suspend fun writeTrainingAttempt(
        questions: List<AttemptQuestion>,
        startedAtMs: Long,
        finishedAtMs: Long,
        topicIds: List<String>,
        extraMeta: JsonObject = JsonObject(emptyMap()),
    ) {
        val total = questions.size
        val correct = questions.count { it.correct == true }
        val durationMs = (finishedAtMs - startedAtMs).toInt()

        val events = buildJsonArray {
            questions.forEach { q ->
                add(buildJsonObject {
                    put("question_id", q.questionId ?: "")
                    put("topic_id", q.topicId ?: "")
                    put("correct", q.correct ?: false)
                    put("chosen_text", q.chosenText ?: "")
                    put("normalized_text", q.normalizedText ?: "")
                    put("correct_text", q.correctText ?: "")
                    put("time_ms", q.timeMs ?: 0)
                    put("difficulty", q.difficulty ?: 1)
                })
            }
        }

        client.rpcVoid(
            "write_answer_events_v1",
            buildJsonObject {
                put("p_source", "test")
                put("p_attempt_ref", UUID.randomUUID().toString().lowercase())
                put("p_events", events)
                put("p_attempt_started_at", Instant.ofEpochMilli(startedAtMs).toString())
                put("p_attempt_finished_at", Instant.ofEpochMilli(finishedAtMs).toString())
                put("p_attempt_meta", buildJsonObject {
                    put("mode", "test")
                    putJsonArray("topic_ids") { topicIds.forEach { add(it) } }
                    put("total", total)
                    put("correct", correct)
                    put("duration_ms", durationMs)
                    put("avg_ms", if (total > 0) durationMs / total else 0)
                    put("client", "android")
                    extraMeta.forEach { (k, v) -> put(k, v) }
                })
            },
        )
    }

    // MARK: Consent (ученик)

    suspend fun incomingTeacherRequests(): List<IncomingTeacherRequest> =
        client.rpc(
            "list_incoming_teacher_requests",
            deserializer = ListSerializer(IncomingTeacherRequest.serializer()),
        )

    suspend fun respondTeacherRequest(requestId: String, accept: Boolean) =
        client.rpcVoid(
            "respond_teacher_request",
            buildJsonObject {
                put("p_request_id", requestId)
                put("p_accept", accept)
            },
        )

    suspend fun myTeachers(): List<MyTeacher> =
        client.rpc("list_my_teachers", deserializer = ListSerializer(MyTeacher.serializer()))

    suspend fun revokeTeacher(teacherId: String) =
        client.rpcVoid("revoke_my_teacher", buildJsonObject { put("p_teacher_id", teacherId) })
}

File: ios/EGETrainerApp/EGETrainerApp/Services/StudentService.swift
import Foundation

/// Студенческая аналитика + consent-flow ученика + запись попыток тренировки.
struct StudentService {
    let client: SupabaseClient

    init(client: SupabaseClient = .shared) {
        self.client = client
    }

    /// student_analytics_screen_v1 — canonical layer-4 контракт статистики.
    /// scope: "self" (ученик о себе) | "teacher" (учитель об ученике, требуется studentId).
    func analytics(
        scope: String = "self",
        studentId: String? = nil,
        days: Int = 30,
        source: String = "all"
    ) async throws -> AnalyticsScreen {
        var params: [String: JSONValue] = [
            "p_viewer_scope": .string(scope),
            "p_days": .number(Double(days)),
            "p_source": .string(source),
            "p_mode": .string("init"),
        ]
        if let studentId {
            params["p_student_id"] = .string(studentId)
        }
        return try await client.rpc("student_analytics_screen_v1", params: params, as: AnalyticsScreen.self)
    }

    /// id текущего пользователя — для self-гейта teacher RPC (как readSessionFallback().user.id).
    func selfUserId() async -> String? {
        await client.currentSession?.user.id
    }

    /// Подбор задач с фильтром — self-гейт teacher_picking_resolve_batch_v1
    /// (тот же RPC, p_student_id = свой id; как student-ветки picker.js).
    /// requests: (scope_kind 'topic'|'section', scope_id, n) — n уже с over-fetch.
    func resolveFiltered(
        requests: [(scopeKind: String, scopeId: String, n: Int)],
        filterId: String?,
        excludeQuestionIds: [String],
        seed: String?
    ) async throws -> [ResolvedQuestion] {
        guard let sid = await selfUserId() else { throw SupabaseError.authRequired }
        let reqs: [JSONValue] = requests.map {
            .object([
                "scope_kind": .string($0.scopeKind),
                "scope_id": .string($0.scopeId),
                "n": .number(Double($0.n)),
            ])
        }
        let result = try await client.rpc("teacher_picking_resolve_batch_v1", params: [
            "p_student_id": .string(sid),
            "p_source": .string("all"),
            "p_filter_id": filterId.map { .string($0) } ?? .null,
            "p_selection": .object([:]),
            "p_requests": .array(reqs),
            "p_seed": seed.map { .string($0) } ?? .null,
            "p_exclude_question_ids": .array(excludeQuestionIds.map { .string($0) }),
        ], as: ResolveBatchResult.self)
        return result.pickedQuestions ?? []
    }

    /// Экран подбора с фильтром для самого ученика (self-гейт
    /// teacher_picking_screen_v2 — бейджи состояний тем как у учителя).
    func pickingScreenSelf(filterId: String?, days: Int = 30) async throws -> PickingScreen {
        guard let sid = await selfUserId() else { throw SupabaseError.authRequired }
        return try await client.rpc("teacher_picking_screen_v2", params: [
            "p_student_id": .string(sid),
            "p_mode": .string("init"),
            "p_days": .number(Double(days)),
            "p_source": .string("all"),
            "p_filter_id": filterId.map { .string($0) } ?? .null,
            "p_selection": .object([:]),
            "p_request": .object([:]),
            "p_seed": .null,
            "p_exclude_question_ids": .null,
        ], as: PickingScreen.self)
    }

    // MARK: - Тренировка: запись попытки (write_answer_events_v1, supabase-write.js)

    func writeTrainingAttempt(
        questions: [AttemptQuestion],
        startedAt: Date,
        finishedAt: Date,
        topicIds: [String],
        extraMeta: [String: JSONValue] = [:]
    ) async throws {
        let events: [JSONValue] = questions.map { q in
            .object([
                "question_id": .string(q.questionId ?? ""),
                "topic_id": .string(q.topicId ?? ""),
                "correct": .bool(q.correct ?? false),
                "chosen_text": .string(q.chosenText ?? ""),
                "normalized_text": .string(q.normalizedText ?? ""),
                "correct_text": .string(q.correctText ?? ""),
                "time_ms": .number(Double(q.timeMs ?? 0)),
                "difficulty": .number(Double(q.difficulty ?? 1)),
            ])
        }
        let total = questions.count
        let correct = questions.filter { $0.correct == true }.count
        let durationMs = Int(finishedAt.timeIntervalSince(startedAt) * 1000)
        let iso = ISO8601DateFormatter()

        var meta: [String: JSONValue] = [
            "mode": .string("test"),
            "topic_ids": .array(topicIds.map { .string($0) }),
            "total": .number(Double(total)),
            "correct": .number(Double(correct)),
            "duration_ms": .number(Double(durationMs)),
            "avg_ms": .number(Double(total > 0 ? durationMs / total : 0)),
            "client": .string("ios"),
        ]
        for (k, v) in extraMeta { meta[k] = v }

        try await client.rpcVoid("write_answer_events_v1", params: [
            "p_source": .string("test"),
            "p_attempt_ref": .string(UUID().uuidString.lowercased()),
            "p_events": .array(events),
            "p_attempt_started_at": .string(iso.string(from: startedAt)),
            "p_attempt_finished_at": .string(iso.string(from: finishedAt)),
            "p_attempt_meta": .object(meta),
        ])
        // WIOS.2: состояние ученика изменилось → витрина локального подбора устарела.
        await PickSnapshotCache.shared.invalidateAll()
        await AccordionScreenCache.shared.invalidateAll()
    }

    // MARK: - Consent (ученик)

    func incomingTeacherRequests() async throws -> [IncomingTeacherRequest] {
        try await client.rpc("list_incoming_teacher_requests", params: [:], as: [IncomingTeacherRequest].self)
    }

    func respondTeacherRequest(requestId: String, accept: Bool) async throws {
        try await client.rpcVoid("respond_teacher_request", params: [
            "p_request_id": .string(requestId),
            "p_accept": .bool(accept),
        ])
    }

    func myTeachers() async throws -> [MyTeacher] {
        try await client.rpc("list_my_teachers", params: [:], as: [MyTeacher].self)
    }

    func revokeTeacher(teacherId: String) async throws {
        try await client.rpcVoid("revoke_my_teacher", params: ["p_teacher_id": .string(teacherId)])
    }
}
